File: kakao-ios-sdk-1.0.64/samples/objc/LinkSample/LinkSample/ViewController.m
/**
 * Copyright 2015 Kakao Corp.
 *
 * Redistribution and modification in source or binary forms are not permitted without specific prior written permission.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "ViewController.h"
#import "IconTableViewCell.h"
#import "KakaoLinkActivity.h"
#import <KakaoOpenSDK/KakaoOpenSDK.h>

#import "StoryLinkHelper.h"
#import "StoryPostingActivity.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource, UIDocumentInteractionControllerDelegate, UIAlertViewDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ViewController {
    NSArray *_menuItems;
    NSArray *_menuItemSubTexts;
    UIDocumentInteractionController *_documentController;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.tableView.tableFooterView = [[UIView alloc] init];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    _menuItems = @[@"Send Link", @"Send Link", @"Share File", @"Story Posting"];
    _menuItemSubTexts = @[@"", @"(UIActivityViewController)", @"(UIDocumentInteractionController)", @""];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSArray *)dummyLinkObjects {
    KakaoTalkLinkObject *label = [KakaoTalkLinkObject createLabel:@"Test Label\nNext Line"];
    KakaoTalkLinkObject *image = [KakaoTalkLinkObject createImage:@"https://developers.kakao.com/assets/img/link_sample.jpg" width:138 height:80];
    KakaoTalkLinkObject *webLink = [KakaoTalkLinkObject createWebLink:@"Test Link" url:@"http://www.kakao.com"];
    
    KakaoTalkLinkAction *androidAppAction = [KakaoTalkLinkAction createAppAction:KakaoTalkLinkActionOSPlatformAndroid devicetype:KakaoTalkLinkActionDeviceTypePhone execparam:@{@"test1" : @"test1", @"test2" : @"test2"}];
    KakaoTalkLinkAction *iphoneAppAction = [KakaoTalkLinkAction createAppAction:KakaoTalkLinkActionOSPlatformIOS devicetype:KakaoTalkLinkActionDeviceTypePhone execparam:@{@"test1" : @"test1", @"test2" : @"test2"}];
    KakaoTalkLinkAction *ipadAppAction = [KakaoTalkLinkAction createAppAction:KakaoTalkLinkActionOSPlatformIOS devicetype:KakaoTalkLinkActionDeviceTypePad execparam:@{@"test1" : @"test1", @"test2" : @"test2"}];
    KakaoTalkLinkObject *appLink = [KakaoTalkLinkObject createAppButton:@"Test Button" actions:@[androidAppAction, iphoneAppAction, ipadAppAction]];

    return @[label, image, webLink, appLink];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 60;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _menuItems.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"IconTableViewCell"];
    if (cell == nil) {
        cell = [[IconTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"IconTableViewCell"];
    }
    
    cell.imageView.image = [UIImage imageNamed:[NSString stringWithFormat:@"LinkIcon%d", (int) indexPath.row]];
    cell.textLabel.text = _menuItems[indexPath.row];
    cell.detailTextLabel.text = _menuItemSubTexts[indexPath.row];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    switch (indexPath.row) {
        case 0:
            [self sendLink];
            break;
        case 1:
            [self showActivity];
            break;
        case 2:
            [self showChooseSharingFile];
            break;
        case 3:
            [self postStory];
            break;
        default:
            break;
    }
}

- (void)sendLink {
    if ([KOAppCall canOpenKakaoTalkAppLink]) {
        [KOAppCall openKakaoTalkAppLink:[self dummyLinkObjects]];
    } else {
        NSLog(@"Cannot open kakaotalk.");
    }
}

- (void)showActivity {
    UIActivity *activity = [[KakaoLinkActivity alloc] initWithObjects:[self dummyLinkObjects]];
    UIActivity *storyActivity = [[StoryPostingActivity alloc] initWithLinkURLString:[self dummyStoryLinkURLString]];
    
    UIActivityViewController *activityViewController = [[UIActivityViewController alloc] initWithActivityItems:@[activity.activityTitle, storyActivity.activityTitle]
                                                                                         applicationActivities:@[activity, storyActivity]];
    
    [self presentViewController:activityViewController animated:YES completion:nil];
    
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad
        && [activityViewController respondsToSelector:@selector(popoverPresentationController)]) {
        activityViewController.popoverPresentationController.permittedArrowDirections = UIPopoverArrowDirectionAny;
        activityViewController.popoverPresentationController.sourceView = self.view;
        activityViewController.popoverPresentationController.sourceRect = CGRectMake(self.view.frame.size.width/2, self.view.frame.size.height/2, 0, 0);
    }
}

- (void)showChooseSharingFile {
    UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"" message:@"공유 파일?" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"JPG", @"MP4", @"TXT", @"GIF", nil];
    [alertView show];
    
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    switch (buttonIndex) {
        case 1:
            [self shareFile:[[NSBundle mainBundle] URLForResource:@"test_img" withExtension:@"jpg"]];
            break;
        case 2:
            [self shareFile:[[NSBundle mainBundle] URLForResource:@"test_vod" withExtension:@"mp4"]];
            break;
        case 3:
            // kakaotalk not support yet.
            [self shareFile:[[NSBundle mainBundle] URLForResource:@"test_text" withExtension:@"txt"]];
            break;
        case 4:
            // kakaotalk not support yet.
            [self shareFile:[[NSBundle mainBundle] URLForResource:@"test_gif" withExtension:@"gif"]];
            break;
        default:
            break;
    }
}

- (void)shareFile:(NSURL *)localPath {
    _documentController = [[UIDocumentInteractionController alloc] init];
    _documentController.URL = localPath;
    _documentController.delegate = self;
    [_documentController presentOptionsMenuFromRect:self.view.frame inView:self.view animated:YES];
}

- (void)documentInteractionControllerDidDismissOptionsMenu:(UIDocumentInteractionController *)controller {
    _documentController = nil;
}

- (NSString *)dummyStoryLinkURLString {
    NSBundle *bundle = [NSBundle mainBundle];
    ScrapInfo *scrapInfo = [[ScrapInfo alloc] init];
    scrapInfo.title = @"Sample";
    scrapInfo.desc = @"Sample 입니다.";
    scrapInfo.imageURLs = @[@"http://www.daumkakao.com/images/operating/temp_mov.jpg"];
    scrapInfo.type = ScrapTypeVideo;
    
    return [StoryLinkHelper makeStoryLinkWithPostingText:@"Sample Story Posting https://www.youtube.com/watch?v=XUX1jtTKkKs"
                                             appBundleID:[bundle bundleIdentifier]
                                              appVersion:[bundle objectForInfoDictionaryKey:@"CFBundleShortVersionString"]
                                                 appName:[bundle objectForInfoDictionaryKey:@"CFBundleName"]
                                               scrapInfo:scrapInfo];
}

- (void)postStory {
    if (![StoryLinkHelper canOpenStoryLink]) {
        NSLog(@"Cannot open kakao story.");
        return;
    }
    
    [StoryLinkHelper openStoryLinkWithURLString:[self dummyStoryLinkURLString]];
}


@end

File: kakao-ios-sdk-1.0.64/samples/swift/LinkSample/LinkSample/ViewController.swift
/**
* Copyright 2015 Kakao Corp.
*
* Redistribution and modification in source or binary forms are not permitted without specific prior written permission.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, UIDocumentInteractionControllerDelegate, UIAlertViewDelegate {

    @IBOutlet weak var tableView: UITableView!
    fileprivate var menuItems: [String] = ["Send Link", "Send Link", "Share File", "Story Posting"]
    fileprivate var menuItemSubTexts: [String] = ["", "(UIActivityViewController)", "(UIDocumentInteractionController)", ""]
    
    fileprivate var documentController: UIDocumentInteractionController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        tableView.tableFooterView = UIView(frame: CGRect.zero)
        tableView.delegate = self
        tableView.dataSource = self
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()

        // Dispose of any resources that can be recreated.
    }
    
    func dummyLinkObject() -> [KakaoTalkLinkObject] {
        let label = KakaoTalkLinkObject.createLabel("Test Label\nNext Line")
        let image = KakaoTalkLinkObject.createImage("https://developers.kakao.com/assets/img/link_sample.jpg", width: 138, height: 80)
        let webLink = KakaoTalkLinkObject.createWebLink("Test Link", url: "http://www.kakao.com")
        
        let androidAppAction = KakaoTalkLinkAction.createAppAction(KakaoTalkLinkActionOSPlatform.android, devicetype: KakaoTalkLinkActionDeviceType.phone, execparam: ["test1" : "test1", "test2" : "test2"])
        let iphoneAppAction = KakaoTalkLinkAction.createAppAction(KakaoTalkLinkActionOSPlatform.IOS, devicetype: KakaoTalkLinkActionDeviceType.phone, execparam: ["test1" : "test1", "test2" : "test2"])
        let ipadAppAction = KakaoTalkLinkAction.createAppAction(KakaoTalkLinkActionOSPlatform.IOS, devicetype: KakaoTalkLinkActionDeviceType.pad, execparam: ["test1" : "test1", "test2" : "test2"])
        let appLink = KakaoTalkLinkObject.createAppButton("Test Button", actions: [androidAppAction!, iphoneAppAction!, ipadAppAction!])
        
        return [label!, image!, webLink!, appLink!]
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 60
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return menuItems.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var normalCell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "Cell")
        if normalCell == nil {
            normalCell = IconTableViewCell(style: UITableViewCellStyle.subtitle, reuseIdentifier: "Cell")
        }
        
        normalCell?.imageView?.image = UIImage(named: "LinkIcon\((indexPath as NSIndexPath).row)")
        normalCell?.textLabel?.text = menuItems[(indexPath as NSIndexPath).row]
        normalCell?.detailTextLabel?.text = menuItemSubTexts[(indexPath as NSIndexPath).row]
        
        return normalCell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        switch (indexPath as NSIndexPath).row {
        case 0:
            sendLink()
        case 1:
            showActivity()
        case 2:
            showChooseSharingFile()
        case 3:
            postStory()
        default:
            fatalError("no menu items.")
        }
    }
    
    func sendLink() {
        if KOAppCall.canOpenKakaoTalkAppLink() {
            KOAppCall.openKakaoTalkAppLink(dummyLinkObject())
        } else {
            print("Cannot open kakaotalk.")
        }
    }
    
    func showActivity() {
        let applicationActivity = KakaoLinkActivity(objects: dummyLinkObject())
        let storyPostingActivity = StoryPostingActivity(linkUrlString: dummyStoryLinkURLString())
        
        let activityViewController = UIActivityViewController(activityItems: [applicationActivity.ACTIVITY_NAME, storyPostingActivity.ACTIVITY_NAME],
            applicationActivities: [applicationActivity, storyPostingActivity])

        present(activityViewController, animated: true, completion: nil)
        
        if UIDevice.current.userInterfaceIdiom == UIUserInterfaceIdiom.pad {
            if #available(iOS 8.0, *) {
                activityViewController.popoverPresentationController!.permittedArrowDirections = UIPopoverArrowDirection.any
                activityViewController.popoverPresentationController!.sourceView = self.view;
                activityViewController.popoverPresentationController!.sourceRect = CGRect(x: self.view.frame.size.width/2, y: self.view.frame.size.height/2, width: 0, height: 0);
            }
        }
    }
    
    func showChooseSharingFile() {
        UIAlertView(title: "", message: "공유 파일?", delegate: self, cancelButtonTitle: "Cancel", otherButtonTitles: "JPG", "MP4", "TXT", "GIF").show()
    }
    
    func alertView(_ alertView: UIAlertView, clickedButtonAt buttonIndex: Int) {
        switch buttonIndex {
        case 0:
            return // cancel
        case 1:
            shareFile(Bundle.main.url(forResource: "test_img", withExtension: "jpg"))
        case 2:
            shareFile(Bundle.main.url(forResource: "test_vod", withExtension: "mp4"))
        case 3:
            shareFile(Bundle.main.url(forResource: "test_text", withExtension: "txt")) // kakaotalk not support yet.
        case 4:
            shareFile(Bundle.main.url(forResource: "test_gif", withExtension: "gif")) // kakaotalk not support yet.
        default:
            fatalError("no item to share.")
        }
    }
    
    func shareFile(_ localPath: URL?) {
        if let localPath = localPath {
            documentController = UIDocumentInteractionController(url: localPath)
            documentController?.delegate = self
            documentController?.presentOptionsMenu(from: self.view.frame, in: self.view, animated: true)
        }
    }
    
    func documentInteractionControllerDidDismissOptionsMenu(_ controller: UIDocumentInteractionController) {
        self.documentController = nil
    }
    
    func dummyStoryLinkURLString() -> String! {
        let bundle = Bundle.main
        var scrapInfo = ScrapInfo()
        scrapInfo.title = "Sample"
        scrapInfo.desc = "Sample 입니다."
        scrapInfo.imageUrls = ["http://www.daumkakao.com/images/operating/temp_mov.jpg"]
        scrapInfo.type = ScrapType.Video
        
        if let bundleId = bundle.bundleIdentifier, let appVersion: String = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String,
            let appName: String = bundle.object(forInfoDictionaryKey: "CFBundleName") as? String {
                
            return StoryLinkHelper.makeStoryLink("Sample Story Posting https://www.youtube.com/watch?v=XUX1jtTKkKs",
                appBundleId: bundleId, appVersion: appVersion, appName: appName, scrapInfo: scrapInfo)
        }
        
        return nil;
    }
    
    func postStory() {
        if !StoryLinkHelper.canOpenStoryLink() {
            print("Cannot open kakao story.")
            return
        }
        
        if let urlString = dummyStoryLinkURLString() {
            _ = StoryLinkHelper.openStoryLink(urlString)
        }
    }
}

